### PR TITLE
Clawback resync

### PR DESCRIPTION
**Pull request by ytx1991**

### CODE REVIEW

File: chia/wallet/wallet_state_manager.py
Comment (xdustinface) [inline chia/wallet/wallet_state_manager.py:1134]: Re-fetching the coin we are currently processing doesn't look right. We would we need to do that?

Comment (ytx1991) [reply chia/wallet/wallet_state_manager.py:1134]: This is for fixing the resync issue of clawback. After the resync, the wallet may not update the clawback coin to spent, which causes the related clawback tx also be stuck in pending status. The possible root cause is the concurrent syncing. We may process or fetch the higher blocks before the block which confirms the clawback coin. To solve this, I added an additional check to fetch the latest coin state. The passed-in coin state is for a certain height.
This fix is verified on the testnet.

Comment (xdustinface) [reply chia/wallet/wallet_state_manager.py:1134]: But why do you need to fetch the coins state again here? Wouldn't it be enough to just do what you do below with `coin_state` instead of `latest_coin_state`? If not, i think it would be good to investigate what is really happening here in the failure case and maybe try to reproduce this issue in a test?

**Comment by xdustinface** [comment]
Reopening because CI obviously only run conflicts checks here.

File: chia/wallet/wallet_state_manager.py
Comment (ytx1991) [reply chia/wallet/wallet_state_manager.py:1134]: Since the coin_state is always unspent, the coin may already have been spent long time ago. This is a general issue for syncing and caused many weird issues, not just for Clawback. We should think about if it is feasible to resync synchronously. I mean only process the higher blocks before all previous blocks are processed.

Comment (ytx1991) [reply chia/wallet/wallet_state_manager.py:1134]: I just added a log. Seems the coin_state has a chance to be set as spent. So the question is why sometimes the coin_State is spent but sometimes it doesn't.

File: chia/wallet/wallet_state_manager.py
Comment (altendky) [inline chia/wallet/wallet_state_manager.py:1156]: This does not address the coverage complaint.  https://github.com/Chia-Network/chia-blockchain/pull/15496/commits/9a2ed50a50e0724768cba358ad26c91c6902bdca  Probably either go back to the original, or take this slightly more explicit option.  Or maybe there's a good third option, I haven't looked too much.

```suggestion
            spent_height: uint32
            if latest_coin_state.spent_height is None:
                spent_height = uint32(0)
            else:
                spent_height = uint32(latest_coin_state.spent_height)
            )
```

**Review by altendky — DISMISSED** [review]

File: chia/wallet/wallet_state_manager.py
Comment (altendky) [reply chia/wallet/wallet_state_manager.py:1156]: If they fix https://github.com/nedbat/coveragepy/issues/509 it would complain regardless.

File: chia/wallet/wallet_state_manager.py
Comment (altendky) [inline chia/wallet/wallet_state_manager.py:1153]: Why should we not test this?

File: chia/wallet/wallet_state_manager.py
Comment (altendky) [reply chia/wallet/wallet_state_manager.py]: Continued at https://github.com/Chia-Network/chia-blockchain/pull/15496#pullrequestreview-1481700019.

File: chia/wallet/wallet_state_manager.py
Comment (ytx1991) [reply chia/wallet/wallet_state_manager.py:1153]: I don't know how to simulate the nuke wallet DB. I extended the resync unit test but it doesn't trigger the change.

**Comment by ytx1991** [comment]
This fix is not necessary for the release, close for now

File: chia/wallet/wallet_state_manager.py
Comment (ytx1991) [reply chia/wallet/wallet_state_manager.py]: This is removed

**Comment by xdustinface** [comment]
@ytx1991 The test you added in 91ac933 doesn't seem to pass. It still running on CI but times out here locally with the following:

```
21:02:20 chia.wallet.wallet_node: INFO Selected multiprocessing start method: default
21:02:20 chia.wallet.wallet_node: INFO Resetting wallet sync data...
21:02:20 chia.wallet.wallet_node: INFO Reset wallet sync data completed.
21:02:20 chia.wallet.wallet_node: DEBUG Trying to disable resync: 2999502625 [2999502625]
21:02:20 chia.wallet.wallet_node: INFO Disabled resync for wallet fingerprint: 2999502625
21:02:20 chia.wallet.wallet_state_manager: DEBUG Starting in db path: /private/var/folders/6b/sb101ygd0_xfxmbc_ljxybr80000gn/T/tmprryrmaxf/test-wallet-db-None-2999502625.sqlite
21:02:20 chia.wallet.wallet_node: INFO Wallet is logged in using key with fingerprint: 2999502625
21:02:20 chia.wallet.wallet_node: INFO Updated last used fingerprint: 2999502625
21:02:20 chia.wallet.did_wallet.did_wallet: INFO Confirmed balance for did wallet is 0
21:02:20 chia.wallet.did_wallet.did_wallet: INFO Confirmed balance for did wallet is 0
FAILED
tests/wallet/rpc/test_wallet_rpc.py:1907 (test_fully_resync[True])
tests/wallet/rpc/test_wallet_rpc.py:1980: in test_fully_resync
    await time_out_assert(20, wc2.get_synced)
E   AssertionError: Timed assertion timed out after 22 seconds: expected True, got False
```

File: chia/wallet/wallet_state_manager.py
Comment (ytx1991) [reply chia/wallet/wallet_state_manager.py:1153]: added test

Comment (trepca) [reply chia/wallet/wallet_state_manager.py:1153]: can we remove the pragma now?

File: chia/wallet/wallet_state_manager.py
Comment (xdustinface) [inline chia/wallet/wallet_state_manager.py:1476]: Seems like a good idea to not use random bytes for a tx hash yeah but its probably not related to the actual problem here?

**Review by xdustinface — CHANGES_REQUESTED** [review]

File: chia/wallet/wallet_state_manager.py
Comment (altendky) [reply chia/wallet/wallet_state_manager.py]: Looks like there are still lines that need covered, but this particular case seems addressed.  Feel free to close this discussion.  I'll dismiss my requested changes.  Thanks.

File: chia/wallet/wallet_state_manager.py
Comment (ytx1991) [reply chia/wallet/wallet_state_manager.py:1163]: In the normal case, if the clawback coin is spent, this handle_clawback will not be executed. It will only be spent here when the wallet first time saw the coin, and it is spent. The tx record shouldn't exist at this point since if it exists the wallet should also have a related coin record. If there is a coin record, this handle_clawback will not be executed.

File: chia/wallet/wallet_state_manager.py
Comment (ytx1991) [reply chia/wallet/wallet_state_manager.py:767]: We can't just drop the first element here since the second one will be the hint which is the actual memo. GUI will hide the first element.

File: chia/wallet/wallet_state_manager.py
Comment (ytx1991) [reply chia/wallet/wallet_state_manager.py:1212]: No, the confirmed has a different meaning here. GUI uses this to know if one clawback tx is claimable.

File: chia/wallet/wallet_state_manager.py
Comment (ytx1991) [reply chia/wallet/wallet_state_manager.py:1170]: They are different. This is an outgoing tx. We user resync wallet, we need to recover it.

File: chia/wallet/wallet_state_manager.py
Comment (ytx1991) [reply chia/wallet/wallet_state_manager.py:1190]: If the confirmed = true, then the wallet will not send the tx again. But I agree to change this.

File: chia/wallet/wallet_state_manager.py
Comment (ytx1991) [reply chia/wallet/wallet_state_manager.py:1476]: This is the original fix for the duplicate issue since we saw duplicate tx records. It seems unrelated now, but good to have. Let me know if you think it's necessary to create another PR for this.

File: tests/wallet/clawback/test_clawback_lifecycle.py
Comment (ytx1991) [reply tests/wallet/clawback/test_clawback_lifecycle.py:201]: We observed a weird transient double-spend error in this test, trying to fix it.

File: chia/wallet/wallet_state_manager.py
Comment (xdustinface) [reply chia/wallet/wallet_state_manager.py]: Why would the GUI hide the first element? Did we introduce this for this clawback case now or is there some general rule we established, that we include empty bytes for the first memo if its not a hint? Im not aware of this. Maybe the GUI should be adjusted instead if this is the only case where this is relevant?

File: chia/wallet/wallet_state_manager.py
Comment (xdustinface) [reply chia/wallet/wallet_state_manager.py:1471]: Hm i will defer to @trepca here but for me it seems like we don't need this in the release then and we can PR it against main instead.

File: tests/wallet/clawback/test_clawback_lifecycle.py
Comment (xdustinface) [reply tests/wallet/clawback/test_clawback_lifecycle.py:201]: Seems like #15619 intends to fix this, right?

File: chia/wallet/wallet_state_manager.py
Comment (xdustinface) [inline chia/wallet/wallet_state_manager.py:1167]: I take it was unused but was it maybe supposed to be passed into the transaction record below rather than being dropped like you did with the other record above?

File: chia/wallet/wallet_state_manager.py
Comment (ytx1991) [reply chia/wallet/wallet_state_manager.py]: We set the first element as b"" majorly because we don't want to mess up others' wallets. If the second element is someone's puzzle hash, the wallet will process this coin but it will throw error since it doesn't own the puzzle hash. The GUI thing exists since the first day I think. Since the first element is set by the wallet, so we don't want to show the wallet generated memo unless it is the only element.

Comment (xdustinface) [reply chia/wallet/wallet_state_manager.py]: > We set the first element as b"" majorly because we don't want to mess up others' wallets.

Can you link me to some other place where we do this? Maybe i just always missed it but i never saw it before.

> If the second element is someone's puzzle hash, the wallet will process this coin but it will throw error since it doesn't own the puzzle hash. 

Why/How would this be the case?

File: tests/wallet/clawback/test_clawback_lifecycle.py
Comment (ytx1991) [reply tests/wallet/clawback/test_clawback_lifecycle.py]: removed

File: chia/wallet/wallet_state_manager.py
Comment (ytx1991) [reply chia/wallet/wallet_state_manager.py:1167]: The problem is this tx can be unconfirmed, and it will resend the spend bundle if it is None.  We will see double spend error in the log.

File: chia/wallet/util/transaction_type.py
Comment (ytx1991) [reply chia/wallet/util/transaction_type.py:18]: It may be because I merged the recent release/1.8.2 and reverted some changes mistakenly. Do you think I should revert this change?

Comment (xdustinface) [reply chia/wallet/util/transaction_type.py:18]: Im fine either way, leaving it in here or dropping since its introduced in the release anyway.

File: chia/wallet/wallet_state_manager.py
Comment (ytx1991) [reply chia/wallet/wallet_state_manager.py]: Everywhere we use hint will have this issues, for examples
For CAT: https://github.com/Chia-Network/chia-blockchain/blob/release/1.8.2/chia/wallet/wallet_state_manager.py#L862
For DID: https://github.com/Chia-Network/chia-blockchain/blob/release/1.8.2/chia/wallet/wallet_state_manager.py#L917
It doesn't mean we add coins to wrong wallets, it just generate some unnecessary error logs in these wallets.

Comment (xdustinface) [reply chia/wallet/wallet_state_manager.py]: My point is, this is not something new and we already now on main in every other case don't know if the first entry, if its a puzzle hash, belongs to us or not so i wonder why should we now in case of clawback start to change the way we handle things usually and introduce a empty first entry? Its seems very unlikely to me that someone puts someone else's puzzle hash into the memo or do i miss something here?

Comment (ytx1991) [reply chia/wallet/wallet_state_manager.py]: For The GUI question, you can try it by sending a CAT with a memo. Every CAT has a hint as the first element and your memo will be the second element. On GUI you will only see the second element.

Comment (xdustinface) [reply chia/wallet/wallet_state_manager.py]: > For The GUI question, you can try it by sending a CAT or Clawback tx with a memo. Every CAT has a hint as the first element and your memo will be the second element. On GUI you will only see the second element.

Okay yeah but a CAT is supposed to have a hint in any case, a clawed/claimed XCH coin not?

File: chia/wallet/wallet_state_manager.py
Comment (ytx1991) [reply chia/wallet/wallet_state_manager.py]: The confirmed means different here. In general, it means "confirmed on the blockchain". but here it means "confirmed the ownership". An alternative solution is adding a new column and index in the tx table to indicate if a tx is claimable.

File: chia/wallet/wallet_state_manager.py
Comment (ytx1991) [reply chia/wallet/wallet_state_manager.py]: As I mentioned above, the empty byte is not for GUI majorly. Assume the memo list is [My_Puzzle_Hash, Your_Puzzle_HASH]. In the common case, your wallet will not process this coin state because the hint is my puzzle hash. But if you drop the first element, then the hint will be your puzzle hash and your wallet will try to process this coin states. But since it doesn't match the actual inner p2 puzzle so it will fail and log an error. In the worst case, it may trigger some unknown bug. So if we want to prevent this unnecessary processing, we need to set the first element to empty or a byte32 that doesn't belong to any wallet.

Comment (xdustinface) [reply chia/wallet/wallet_state_manager.py]: Did you read this https://github.com/Chia-Network/chia-blockchain/pull/15496#discussion_r1242716607?

File: chia/wallet/wallet_state_manager.py
Comment (trepca) [reply chia/wallet/wallet_state_manager.py:1473]: @ytx1991  lets PR against main

File: chia/wallet/wallet_state_manager.py
Comment (trepca) [reply chia/wallet/wallet_state_manager.py]: a comment explaining this would be good

File: chia/wallet/wallet_state_manager.py
Comment (ytx1991) [reply chia/wallet/wallet_state_manager.py]: I agree it's a rare case, I will change this. We'd better merge this before the release.

File: chia/wallet/wallet_state_manager.py
Comment (ytx1991) [reply chia/wallet/wallet_state_manager.py]: Added

File: .github/workflows/test.yml
Comment (xdustinface) [inline .github/workflows/test.yml:44]: This shouldn't be here.

File: chia/wallet/wallet_state_manager.py
Comment (xdustinface) [inline chia/wallet/wallet_state_manager.py:1211]: Seems like this comment should sit next to the `TransactionRecord` creation and not above the timestamp.
```suggestion
            # We use TransactionRecord.confirmed to indicate if a Clawback transaction is claimable
```

File: tests/wallet/rpc/test_wallet_rpc.py
Comment (xdustinface) [inline tests/wallet/rpc/test_wallet_rpc.py:1998]: Why was this moved?

**Review by xdustinface — CHANGES_REQUESTED** [review]
There is still a leftover from the retargeting to `main` and two nits.

File: tests/wallet/rpc/test_wallet_rpc.py
Comment (ytx1991) [reply tests/wallet/rpc/test_wallet_rpc.py:1998]: I didn't change this. it should be because the base branch changed. I can change it back

File: chia/wallet/wallet_state_manager.py
Comment (xdustinface) [inline chia/wallet/wallet_state_manager.py:765]:
```suggestion
                memos: List[bytes] = incoming_tx.memos[0][1][1:]
```

Is there any scenario how the memos could be empty here?

**Review by xdustinface — DISMISSED** [review]

**Review by danieljperry — COMMENTED** [review]
I cloned `main` and added this PR. I'm having trouble getting auto claim to work. This video shows when it fails (I did cut out a lot where nothing was happening, but I didn't cut the auto claim part):


https://github.com/Chia-Network/chia-blockchain/assets/69756004/f1a33a3e-0367-4aa4-a429-08e443ae7145

Here's my log from when the claim happened:

[auto_claim.txt](https://github.com/Chia-Network/chia-blockchain/files/11908780/auto_claim.txt)

**Comment by danieljperry** [comment]
Here's another issue I saw while attempting to claw back a transaction. Note that this problem might not be related to this PR:


https://github.com/Chia-Network/chia-blockchain/assets/69756004/58132ca6-e735-47b8-b11b-069a81a4fd3f

Log snippet:

```2023-06-29T11:34:13.924 wallet chia.rpc.wallet_rpc_api    : ERROR    Failed to get transaction 1b332cee62521ffe34d04712db291eeb3681e2453ff50811d65d0b87c9d470c3.
Traceback (most recent call last):
  File "C:\Users\User\Chia\main+15496\chia-blockchain\chia\rpc\wallet_rpc_api.py", line 912, in get_transactions
    assert record is not None, f"Cannot find coin record for clawback transaction {tx['name']}"
AssertionError: Cannot find coin record for clawback transaction 0x1b332cee62521ffe34d04712db291eeb3681e2453ff50811d65d0b87c9d470c3
2023-06-29T11:34:22.696 wallet chia.rpc.rpc_server        : WARNING  Error while handling message: Traceback (most recent call last):
  File "C:\Users\User\Chia\main+15496\chia-blockchain\chia\rpc\rpc_server.py", line 340, in safe_handle
    response = await self.ws_api(message)
  File "C:\Users\User\Chia\main+15496\chia-blockchain\chia\rpc\rpc_server.py", line 331, in ws_api
    return await f_rpc_api(data)
  File "C:\Users\User\Chia\main+15496\chia-blockchain\chia\rpc\wallet_rpc_api.py", line 1062, in spend_clawback_coins
    coin_ids: List[bytes32] = [bytes32.from_hexstr(coin) for coin in request["coin_ids"]]
  File "C:\Users\User\Chia\main+15496\chia-blockchain\chia\rpc\wallet_rpc_api.py", line 1062, in <listcomp>
    coin_ids: List[bytes32] = [bytes32.from_hexstr(coin) for coin in request["coin_ids"]]
  File "C:\Users\User\Chia\main+15496\chia-blockchain\chia\util\byte_types.py", line 51, in from_hexstr
    if input_str.startswith("0x") or input_str.startswith("0X"):
AttributeError: 'NoneType' object has no attribute 'startswith'```

File: chia/wallet/wallet_state_manager.py
Comment (ytx1991) [reply chia/wallet/wallet_state_manager.py:765]: It shouldn't be empty, just in case of unexpected things happen.

**Comment by ytx1991** [comment]
> Here's another issue I saw while attempting to claw back a transaction. Note that this problem might not be related to this PR:
> 
>  nonetype.mp4 
> Log snippet:
> 
> ```
> Traceback (most recent call last):
>   File "C:\Users\User\Chia\main+15496\chia-blockchain\chia\rpc\wallet_rpc_api.py", line 912, in get_transactions
>     assert record is not None, f"Cannot find coin record for clawback transaction {tx['name']}"
> AssertionError: Cannot find coin record for clawback transaction 0x1b332cee62521ffe34d04712db291eeb3681e2453ff50811d65d0b87c9d470c3
> 2023-06-29T11:34:22.696 wallet chia.rpc.rpc_server        : WARNING  Error while handling message: Traceback (most recent call last):
>   File "C:\Users\User\Chia\main+15496\chia-blockchain\chia\rpc\rpc_server.py", line 340, in safe_handle
>     response = await self.ws_api(message)
>   File "C:\Users\User\Chia\main+15496\chia-blockchain\chia\rpc\rpc_server.py", line 331, in ws_api
>     return await f_rpc_api(data)
>   File "C:\Users\User\Chia\main+15496\chia-blockchain\chia\rpc\wallet_rpc_api.py", line 1062, in spend_clawback_coins
>     coin_ids: List[bytes32] = [bytes32.from_hexstr(coin) for coin in request["coin_ids"]]
>   File "C:\Users\User\Chia\main+15496\chia-blockchain\chia\rpc\wallet_rpc_api.py", line 1062, in <listcomp>
>     coin_ids: List[bytes32] = [bytes32.from_hexstr(coin) for coin in request["coin_ids"]]
>   File "C:\Users\User\Chia\main+15496\chia-blockchain\chia\util\byte_types.py", line 51, in from_hexstr
>     if input_str.startswith("0x") or input_str.startswith("0X"):
> AttributeError: 'NoneType' object has no attribute 'startswith'```
> ```

It seems like the clawback coin is missing for some reason. I assume it is related to wallet switching. For the auto claim, did you wait several minutes after the timelock expired?

**Comment by danieljperry** [comment]
> > Here's another issue I saw while attempting to claw back a transaction. Note that this problem might not be related to this PR:
> > nonetype.mp4
> > Log snippet:
> > ```
> > Traceback (most recent call last):
> >   File "C:\Users\User\Chia\main+15496\chia-blockchain\chia\rpc\wallet_rpc_api.py", line 912, in get_transactions
> >     assert record is not None, f"Cannot find coin record for clawback transaction {tx['name']}"
> > AssertionError: Cannot find coin record for clawback transaction 0x1b332cee62521ffe34d04712db291eeb3681e2453ff50811d65d0b87c9d470c3
> > 2023-06-29T11:34:22.696 wallet chia.rpc.rpc_server        : WARNING  Error while handling message: Traceback (most recent call last):
> >   File "C:\Users\User\Chia\main+15496\chia-blockchain\chia\rpc\rpc_server.py", line 340, in safe_handle
> >     response = await self.ws_api(message)
> >   File "C:\Users\User\Chia\main+15496\chia-blockchain\chia\rpc\rpc_server.py", line 331, in ws_api
> >     return await f_rpc_api(data)
> >   File "C:\Users\User\Chia\main+15496\chia-blockchain\chia\rpc\wallet_rpc_api.py", line 1062, in spend_clawback_coins
> >     coin_ids: List[bytes32] = [bytes32.from_hexstr(coin) for coin in request["coin_ids"]]
> >   File "C:\Users\User\Chia\main+15496\chia-blockchain\chia\rpc\wallet_rpc_api.py", line 1062, in <listcomp>
> >     coin_ids: List[bytes32] = [bytes32.from_hexstr(coin) for coin in request["coin_ids"]]
> >   File "C:\Users\User\Chia\main+15496\chia-blockchain\chia\util\byte_types.py", line 51, in from_hexstr
> >     if input_str.startswith("0x") or input_str.startswith("0X"):
> > AttributeError: 'NoneType' object has no attribute 'startswith'```
> > ```
> 
> It seems like the clawback coin is missing for some reason. I assume it is related to wallet switching. For the auto claim, did you wait several minutes after the timelock expired?

In this case for autoclaim, I switched to the receiver's wallet before the timelock expired. The autoclaim was then attempted in that wallet automatically as soon as the timelock expired.

**Review by wjblanke — APPROVED** [review]
aok

**Pull request merged**
